Add a guide for distro integrators (#5)
aarch64-laptops/aarch64-laptops.github.io: distro_integration.md

---
title: Distro Integration
layout: page
---

This is a compation guide for distro integrators who want to add support for
aarch64 (particularly Snapdragon X1) devices. This is not a guide for end-users
and assumes that you are already familiar with your distros process for building

> kernels, the initramfs, and options to override the devicetree selection. It

Reviewer: ```suggestion
kernels, the initramfs and options to override the devicetree selection. It
```

Reviewer 2: it's an oxford comma i thought?

Reviewer: I'm most definitely against the Oxford comma, which is more used in American English text. We are an English company ;-)

should be used in conjunction with your distros documentation. The primary
audience is folks involved in distro development but who have either never
worked on device bringup, or never for ARM devices. Feel free to skim/skip over

> sections which don't apply to you.

Reviewer: `Feel free to skim/skip over sections which don't apply to you.` seems quite patronising to me and not sure what benefit it brings. In most documents we have the free ability to skip over sections :-)

Reviewer 2: yeah i'll rework this... thanks

It is recommended that you read through the whole guide before attempting

> bringup, and then use it as a reference while you do. This will help you get a

Reviewer: ```suggestion
bringup and then use it as a reference while you do. This will help you get a
```

better understanding of the whole picture and how each part of the process fits
in.

## Background

Why is it so much harder to get distros up and running on ARM laptops?
**Especially** considering the (seemingly) good support in the Linux kernel?

This is a complicated question, but largely boils down to a few things which are
worth considering as your work to add ARM laptop support to your distro:

1. Devicetree (the hardware descriptor used instead of ACPI for most ARM

>    platforms) is much lower level, and does not contain arbitrary logic. Linux

Reviewer: ```suggestion
   platforms) is much lower level and does not contain arbitrary logic. Linux
```

   must be aware of much lower level details of the hardware. Additionally,
   there are still growing pains with devicetree and missing features that make
   it undesirable to separate the devicetree and kernel versions.

> 2. Historically ARM has been almost entirely in embedded devices which have

Reviewer: ```suggestion
2. Historically, ARM has been almost entirely in embedded devices which have
```

   extremely tight vertical integration. There is a lot of momentum in this
   direction and as a result much of the firmware interfaces, architecture
   designs and software practises lean into this expectation that the

>    devicetree, kernel, and userspace will all be tied together tightly. This

Reviewer: ```suggestion
   devicetree, kernel and userspace will all be tied together tightly. This
```

Reviewer 2: Also an oxford comma?

            gpios = <&pm8998_gpios 6 GPIO_ACTIVE_LOW>;
            debounce-interval = <15>;
        };
    };
}
```

This is really just scratching the surface, but this isn't a guide for writing
devicetrees. If you're interested, check out these resources:

* [eLinux Devicetree Usage](https://elinux.org/Device_Tree_Usage)
* [A whole lot of resources about
  DT](https://elinux.org/Device_Tree_presentations_papers_articles)

The only other significant information to know about devicetrees, is that the
de-facto source for them is in the kernel itself. While they are meant to be
forwards AND backwards compatible with the kernel, this is not always the case.
It's also common that drivers and associated DT are added at the same time,
meaning that if you have an old devicetree that predates displayport alt-mode
support, it's unlikely that you'll be able to use the feature even on a newer
kernel until you update your DT.

### Drivers

Expanding on the point of devicetree being lower level than ACPI (and ARM
platforms generally requiring tighter kernel integration); on an ACPI system
when the kernel wants to power up e.g. a USB controller, it uses the ACPI to do

> it, and usually doesn't need to consider low level details like what power rails

Reviewer: ```suggestion
it and usually doesn't need to consider low level details like what power rails
```

---
title: Distro Integration
layout: page
---

> This is a compation guide for distro integrators who want to add support for

Reviewer: ```suggestion
This is a companion guide for distro integrators who want to add support for
```

the scope of this guide.

## Part 3 - Kernel modules and the initramfs

The eternal battle of every device enablement is getting things to work properly
in the initramfs. This is such a hated process that we have mostly given up and
just run udevd or even systemd itself in the ramdisk. However one thing we still
don't do (most of the time anyway) is pull in all the kernel modules.

On typical x86/ACPI devices, the list of modules required for hardware support
is usually quite generic and mkinitfs implementations have gained the ability
to automatically figure out what modules are needed for the device you're
booting on. But - as is the running theme here - ARM devices usually require way
more modules for basic stuff like display output and keyboard input (the bare
necessities for full disk encryption!).

The author of this guide recommends that you make your life as simple as
possible by doing the following:

1. Install **ALL** kernel modules in the initramfs, especially for initial
   bringup while you're tracking down all the kernel config options.
2. Once your system is up and running with all the expected functionality
   working, you can use [the script

>    here](https://gist.github.com/calebccff/f5e33a25c56c1b2bf32c199bdb76e0de) to

Reviewer 3: You want to keep the script in the gist or have it together with the doc here?

Reviewer 2: Not sure, the gist makes it easier to update later on (and folks can leave comments). But is ofc more brittle.

Reviewer 3: To be fair I don't mind either way. As long as the gist is being kept alive it has the advantage for updating and interaction). Just seen to often that links went into Nirvana after some time.

Reviewer 2: OK, I added a copy of the script to this repo so it's always archived, but kept the gist as the primary link for now, maybe in the future we either just use it from this repo or some other "utils" repo elsewhere...

* [A whole lot of resources about
  DT](https://elinux.org/Device_Tree_presentations_papers_articles)

The only other significant information to know about devicetrees, is that the
de-facto source for them is in the kernel itself. While they are meant to be
forwards AND backwards compatible with the kernel, this is not always the case.
It's also common that drivers and associated DT are added at the same time,
meaning that if you have an old devicetree that predates displayport alt-mode
support, it's unlikely that you'll be able to use the feature even on a newer
kernel until you update your DT.

### Drivers

Expanding on the point of devicetree being lower level than ACPI (and ARM
platforms generally requiring tighter kernel integration); on an ACPI system
when the kernel wants to power up e.g. a USB controller, it uses the ACPI to do
it, and usually doesn't need to consider low level details like what power rails
must be turned on. On devicetree systems, the kernel needs to model all of these
power rails and control them itself.

Drivers for things like "usb repeaters" (chips which help reduce signal
degredation across the mainboard) aren't needed on ACPI - the ACPI itself can

> program these chips, or offload it to the embedded controller. On Snapdragon

Reviewer: ```suggestion
program these chips or offload it to the embedded controller. On Snapdragon
```

laptops this is not usually the case, we need to model the repeater in
devicetree AND write a driver for it (and wire that driver up to properly
receive events).

> The consequence of all of this, is that you need to enable the USB repeater

Reviewer: ```suggestion
The consequence of all of this is that you need to enable the USB repeater
```

> driver in your kernel config, and if you want the USB port to work correctly in

Reviewer: ```suggestion
driver in your kernel config and if you want the USB port to work correctly in
```

Reviewer 2: maybe the comma is wrong but there needs to be a pause here i think

Reviewer: How about

The consequence of all of this is that you need to enable the USB repeater
driver in your kernel config: if you want the USB port to work correctly in

the initramfs (e.g. because you're booting the OS installer from a USB drive)
then the module must be included in the ramdisk.

If you don't do this correctly, you /should/ get some clear error messages from
the kernel, and be able to inspect `/sys/kernel/debug/devices_deferred` to
determine what's missing, however there are many situations where missing
components like this aren't so easy to narrow down.

## Part 1 - The missing kernel config

For Snapdragon laptops, the mandate of the upstream kernel developers is to
ensure that the arm64 defconfig (`arch/arm64/configs/defconfig` for those
following along at home) has everything needed to boot on at least the
well-supported laptops (like the ThinkPad T14s and X13s). Though patches
enabling drivers for other laptops (like display panel drivers) would likely be
welcome.

Where possible, it's recommended that your distro kernel picks up drivers that

> are enabled in defconfig, or even derives itself from the defconfig. This will

Reviewer: ```suggestion
are enabled in defconfig or even derives itself from the defconfig. This will
```

ensure that as new devices are enabled upstream they will trickle-down and may
become supported in your distro without any intentional effort.

### Config fragments

A good approach for e.g. a linux-next kernel variant is to use kernel config
fragments, this allows you to maintain a list of config options to enable or
disable on top of the defconfig. An example of this can be seen in the
[postmarketOS linux-next
package](https://gitlab.postmarketos.org/postmarketOS/pmaports/-/tree/082b6f1908db1cf15daacf862e7bfd21d9f2bf57/device/testing/linux-next)

> which uses a fragment for distro options, and another for device options. These

Reviewer: ```suggestion
which uses a fragment for distro options and another for device options. These
```

waste hours of your time. In general if you're feeling stuck don't hesitate to
head over to the `#aarch64-laptops` IRC channel over on OFTC.

In broad strokes, these are the things to look out for:

1. Pick the LATEST kernel! While support is getting better, Snapdragon laptops
   are getting bug fixes and new features with every single kernel release. If
   your distro offers multiple kernel versions, always use the latest.
2. Ensure your DT is being picked up. For development purposes it is totally
   fine to hardcode this in your GRUB or systemd-boot config (though how exactly
   to do this will depend on your distro).
3. Ensure your DT is the one that came with your kernel. If you made a copy of
   your DT (for example moving it to a generic path so that
   [dtbloader](https://github.com/TravMurav/dtbloader/) would see it) and then
   later moved to a newer kernel or made changes to the DT, it can be easy to
   forget and keep using the old one instead.
4. Set `clk_ignore_unused pd_ignore_unused` in the kernel cmdline. Since
   schematics are often not available, some clocks and power domains may be

>    modelled incorrectly, these options ensure that the resources won't be turned

Reviewer: ```suggestion
   modelled incorrectly. These options ensure that the resources won't be turned
```

the right direction and to check that you didn't miss anything from the
defconfig.

It can be run like this:

```sh
./scripts/dtc/dt_to_config \
    --config /path/to/dot-config \
    --exclude-flag H,y,m --include-flag c \
    --short-name \
    arch/arm64/boot/dts/qcom/x1e78100-lenovo-thinkpad-t14s.dts
```

Where `/path/to/dot-config` is the path to a `.config` (this will either be
tracked in your distro packaging or produced as part of the build process).

Running the above will output a list of devicetree nodes, the `compatible`
properties (this is how drivers match against the devicetree) and associated
driver and kernel config option if known.

For example if you see

```txt
-d-c-----n--F : clock-controller@100000 : qcom,x1e80100-gcc : drivers/clk/qcom/gcc-x1e80100.c : CONFIG_CLK_X1E80100_GCC : n
```

This tells you that the `clock-controller@100000` node with compatible
`qcom,x1e80100-gcc` matched to driver `gcc-x1e80100.c` which is enabled by
`CONFIG_CLK_X1E80100_GCC` which has a current value of `n`...

**In other words**, `CONFIG_CLK_X1E80100_GCC` needs to be enabled.

Please note that some generic nodes will produce incorrect and spurious output.

> This tool is NOT to be trusted, and a second source of truth should be used

Reviewer: ```suggestion
This tool is NOT to be trusted and a second source of truth should be used
```
although "not to be trusted" sounds overly terse

```makefile
obj-$(CONFIG_DRM_PANEL_SAMSUNG_ATNA33XC20) += panel-samsung-atna33xc20.o
```

There you go, `CONFIG_DRM_PANEL_SAMSUNG_ATNA33XC20` is what we need to enable for the panel driver in the Yoga Slim 7x.

#### The pesky last few

There's always a few things missing that are a total pain to track down, you can
check if there's anything in `/sys/kernel/debug/devices_deferred`, though note
that missing firmware (a likelehood if you aren't on a ThinkPad) will cause some
drivers not to probe, we'll come back to these later.

Don't be afraid to `grep` for compatible strings in the `drivers/` directory of
the kernel, you can also usually work backwords from a C file, through the
`Makefile` in the same directory and to the associated kernel config option
which you need.

And of course, head over to the IRC where folks can probably help you translate
your error message or missing functionality into the necessary config options.

## Part 2 - Firmware

Many of the benefits of ARM systems come from their heavy use of co-processors

> for tasks like audio processing, compute, video decoding, etc (notably things

Reviewer: ```suggestion
for hardware-accelerated tasks like audio processing, compute or video decoding (notably things
```

that the x86 world has been trying to catch up with in the last few years).

> But these co-processors all need to run firmware, and that firmware is usually

Reviewer: ```suggestion
These co-processors all need to run firmware and that firmware is usually
```

tied to your specific device. Some of these are packaged in
[linux-firmware](https://web.git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/)
but for most devices there is no clear license and such packaging isn't
possible. Instead the firmware must be pulled from the Windows partition (and
dual booting is the only way to get firmware updates). The only "correct" way to

> handle this is to have your users dual-boot, and to periodically boot back into

Reviewer: ```suggestion
handle this is to have your users dual-boot and to periodically boot back into
```

Windows for updates, before then copying over new firmware versions.

For X1E laptops, the

> [woa-firmware-yoinker](https://github.com/CadmiumLinux/woa-firmware-yoinker/blob/main/yoink-firmware)

Reviewer: there is also some project by the ubuntu maintainers (i could find a link if you care about adding it)

Reviewer 2: Would be good to add for sure!

Reviewer: https://salsa.debian.org/debian/qcom-firmware-extract

Reviewer 2: Thanks! Added that too.

script can be used to fetch this firmware for you after installing. The process
is fairly straightforward and can be applied to other Snapdragon laptops too,

> however it requires disabling Windows bitlocker in order to mount the windows

Reviewer: ```suggestion
however it requires disabling BitLocker encryption on the Windows partition in order to mount the Windows
```

partition from Linux.

The best approach today is probably to come up with a manual or semi-manual
process that users can follow to extract the firmware from windows and put it in
the right place during or after the installation.

Solving the current licensing issues of firmware distribution is outside
the scope of this guide.

## Part 3 - Kernel modules and the initramfs

The eternal battle of every device enablement is getting things to work properly
in the initramfs. This is such a hated process that we have mostly given up and
just run udevd or even systemd itself in the ramdisk. However one thing we still
don't do (most of the time anyway) is pull in all the kernel modules.

On typical x86/ACPI devices, the list of modules required for hardware support

> is usually quite generic, and mkinitfs implementations have gained the ability

Reviewer: ```suggestion
is usually quite generic and mkinitfs implementations have gained the ability
```

signed and validated).

### dtbloader

[dtbloader](https://github.com/TravMurav/dtbloader/) is an EFI driver that can
be used in conjunction with systemd-boot, it loads dtbs from a well-known
directory (`/dtbs/` on the ESP and has it's own internal database of HWIDs). It
is currently used by the postmarketOS trailblazer generic ARM64 EFI target.

To use it, you must package it in your distro and install it into the ESP as a
driver for systemd-boot.

## Part 5 - Community buy-in & Conclusion

There are still a lot of pain points for both distro integrators and end-users
on these devices. Things like audio and external displays are still not always
working great (or at all). While there is plenty of work going on in the kernel,
a lot of effort is still needed in distros and middleware to make these devices
properly nice to install and use.

Adding support for new distros is one of the best ways to help with this effort.
Documenting your process (and more generally the state of support for specific
devices) also goes a long way to spread broader community knowledge and
awareness.

As we collectively become more familiar with ARM platforms and begin to solve

> some of the more infuriating issues, more awareness of how they differ to x86,

Reviewer: ```suggestion
some of the more infuriating issues (e.g. more awareness of how they differ to x86,
```

> what a "devicetree" is, etc. will help us continue bringing the entire stack

Reviewer: ```suggestion
what a "devicetree" is) will help us continue bringing the entire stack
```
